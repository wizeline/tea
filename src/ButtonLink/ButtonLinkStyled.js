// @flow

import Link from '../Link/LinkStyled';
import type { ButtonLinkProps } from './ButtonLinkTypes';

const getSVGFill = ({ color, theme }: ButtonLinkProps): string =>
  color || theme.textLink;

const BaseButtonLinkStyled = Link.withComponent('button');

const ButtonLinkStyled = BaseButtonLinkStyled.extend`
  display: inline-flex;
  align-items: center;
  justify-content: center;
  border: none;
  background: transparent;

  & svg {
    fill: ${getSVGFill};
  }
`;

export default ButtonLinkStyled;
