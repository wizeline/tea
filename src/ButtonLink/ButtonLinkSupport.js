// @flow

import { css } from 'styled-components';
import { getTextStyleSupport } from '../Text/TextSupport';
import { supportBooleanNameGroup } from '../utils';
import type { ButtonLinkProps } from './ButtonLinkTypes';

const priorityPropName = 'priority';
const supportedPriorities = ['b50', 'b100', 'b150', 'h300'];
const defaultPriority = 'b100';

const getButtonLinkPropSupport = (props: ButtonLinkProps) => {
  const { children, ...otherProps } = props;
  const priority = supportBooleanNameGroup(
    priorityPropName,
    supportedPriorities,
    defaultPriority,
  )(otherProps);
  const linkAnchorStyledProps = {
    ...priority,
  };
  return {
    children,
    linkAnchorStyledProps,
  };
};

const getButtonLinkStyleSupport = (props: ButtonLinkProps) => {
  const color = props.color ? props.color : props.theme.textLink;
  const colorOverride = `color: ${color};`;
  return css`
    ${getTextStyleSupport};
    ${colorOverride};
  `;
};

export { getButtonLinkPropSupport, getButtonLinkStyleSupport };
