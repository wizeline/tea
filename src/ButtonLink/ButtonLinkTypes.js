// @flow

import * as React from 'react';

export type ButtonLinkProps = {
  b100?: boolean,
  b150?: boolean,
  b50?: boolean,
  children?: React.Node,
  color?: string,
  h300?: boolean,
  href?: string,
  onClick: () => mixed,
  theme: {
    textLink: string,
  },
};
