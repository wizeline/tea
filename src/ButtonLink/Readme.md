#### Supports all the `<Link />` props, but with button behaviour

```jsx
 <div>
   Default: <ButtonLink >Wizeline</ButtonLink>
   <br />B50: <ButtonLink b50 >Wizeline</ButtonLink>
   <br />B100: <ButtonLink b100 >Wizeline</ButtonLink>
   <br />B150: <ButtonLink b150 >Wizeline</ButtonLink>
   <br />H300: <ButtonLink h300 >Wizeline</ButtonLink>
 </div>
```
