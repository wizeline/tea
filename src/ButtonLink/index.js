// @flow

export { default } from './ButtonLink';
