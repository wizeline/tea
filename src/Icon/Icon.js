// @flow

/* eslint flowtype-errors/enforce-min-coverage: 0 */

import React from 'react';
import constants from './IconConstants';
import type { IconProps } from './IconTypes';
import { IconDivStyled, IconSvgStyled } from './IconStyled';
import getSVG from './data';

const Icon = (props: IconProps) => {
  const { name, size } = props;
  const svg = getSVG(name);
  if (!svg) return null;
  const svgViewBox = svg.props.viewBox || constants.defaultViewBox;
  const svgChildren = svg.props.children;
  return (
    <IconDivStyled size={size}>
      <IconSvgStyled
        iconFillColor={props.color}
        viewBox={svgViewBox}
        preserveAspectRatio={constants.preserveAspectRatio}
      >
        {svgChildren}
      </IconSvgStyled>
    </IconDivStyled>
  );
};

export default Icon;
