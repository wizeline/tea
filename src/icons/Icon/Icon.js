// @flow

import React from 'react';
import CircleIcon from '../CircleIcon/CircleIcon';
import SVGIcon from '../SVGIcon/SVGIcon';
import type { IconPropTypes } from './IconPropTypes';
import getChildProps from './IconHelpers';
import getSVG from '../SVGIcon/data';
import { warn } from '../../utils';
import defaultIconTheme from '../../themes/defaultTheme/components/Icon';

const getInvalidIconNameWarningMessage = (name: string) =>
  `An invalid \`name\` prop with value \`${name}\` was supplied to <Icon />.`;

const getSizeProps = ({ size }) => {
  switch (size) {
    case 'xsmall':
      return { xmall: true };
    case 'medium':
      return { medium: true };
    case 'large':
      return { large: true };
    case 'xlarge':
      return { xlarge: true };
    default:
      return { small: true };
  }
};

const Icon = (iconProps: IconPropTypes) => {
  const { name, round } = iconProps;
  const svg = getSVG(name);
  if (!svg) {
    const warningMessage = getInvalidIconNameWarningMessage(name);
    return warn(warningMessage);
  }
  const childProps = { ...getChildProps(iconProps), svgData: svg };
  const sizeProps = getSizeProps(childProps);
  return round ? (
    <CircleIcon {...childProps} {...sizeProps} />
  ) : (
    <SVGIcon {...childProps} {...sizeProps} />
  );
};

Icon.defaultProps = {
  theme: defaultIconTheme,
};

export default Icon;
