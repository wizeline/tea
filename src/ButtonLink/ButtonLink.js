// @flow

import React from 'react';
import defaultLinkTheme from '../themes/defaultTheme/components/Link';
import { noop } from '../utils';
import ButtonLinkStyled from './ButtonLinkStyled';
import { getButtonLinkPropSupport } from './ButtonLinkSupport';
import type { ButtonLinkProps } from './ButtonLinkTypes';

const ButtonLink = (props: ButtonLinkProps) => {
  const { linkAnchorStyledProps } = getButtonLinkPropSupport(props);

  return <ButtonLinkStyled {...linkAnchorStyledProps} {...props} />;
};

ButtonLink.defaultProps = {
  theme: defaultLinkTheme,
  onClick: noop,
};

export default ButtonLink;
