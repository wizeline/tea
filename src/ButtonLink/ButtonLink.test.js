import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { Icon } from '../icons/Icon';
import ButtonLink from './ButtonLink';

describe('ButtonLink', () => {
  it('functions correctly with children', () => {
    const childrenString = 'hello';
    const props = {
      b150: true,
      children: childrenString,
    };
    const wrapper = shallow(<ButtonLink {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('functions correctly without children', () => {
    const props = {
      b150: true,
    };
    const wrapper = shallow(<ButtonLink {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('functions correctly with svg child', () => {
    const props = {
      b150: true,
    };
    const wrapper = shallow(
      <ButtonLink {...props}>
        <Icon name="add-item" xsmall />
      </ButtonLink>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('functions correctly with custom color', () => {
    const props = {
      color: 'red',
    };
    const wrapper = shallow(<ButtonLink {...props} />);
    expect(wrapper).toHaveStyleRule('color', 'red');
  });

  it('calls onClick function', () => {
    const onClickMocked = jest.fn();
    const props = {
      onClick: onClickMocked,
    };
    const buttonLink = shallow(<ButtonLink {...props} />);
    buttonLink.simulate('click');
    expect(onClickMocked.mock.calls.length).toEqual(1);
  });

  it('has default onClick function', () => {
    const props = {};
    const buttonLink = shallow(<ButtonLink {...props} />);
    buttonLink.simulate('click');
    expect(buttonLink.props().onClick).toBeDefined();
  });
});
