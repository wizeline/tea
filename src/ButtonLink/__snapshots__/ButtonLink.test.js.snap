// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonLink functions correctly with children 1`] = `
<styled.button
  b150={true}
  onClick={[Function]}
  priority="b150"
  theme={
    Object {
      "textLink": "#50A5E7",
    }
  }
>
  hello
</styled.button>
`;

exports[`ButtonLink functions correctly with svg child 1`] = `
<styled.button
  b150={true}
  onClick={[Function]}
  priority="b150"
  theme={
    Object {
      "textLink": "#50A5E7",
    }
  }
/>
`;

exports[`ButtonLink functions correctly without children 1`] = `
<styled.button
  b150={true}
  onClick={[Function]}
  priority="b150"
  theme={
    Object {
      "textLink": "#50A5E7",
    }
  }
/>
`;
